cloud-controller-manager/cmd/digitalocean-cloud-controller-manager-admission-server/main.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2023 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/oauth2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	"github.com/digitalocean/digitalocean-cloud-controller-manager/cloud-controller-manager/do"
+	"github.com/digitalocean/godo"
+)
+
+const (
+	doAccessTokenEnv    string = "DO_ACCESS_TOKEN"
+	doOverrideAPIURLEnv        = "DO_OVERRIDE_URL"
+	doClusterIDEnv             = "DO_CLUSTER_ID"
+	doClusterVPCIDEnv          = "DO_CLUSTER_VPC_ID"
+)
+
+var loggerVerbosity = flag.Int("v", 0, "logger verbosity")
+
+func init() {
+	flag.Parse()
+	log.SetLogger(zap.New(zap.Level(zapcore.Level(-*loggerVerbosity))))
+}
+
+func main() {
+	if err := startAdmissionServer(); err != nil {
+		fmt.Fprintf(os.Stderr, "failed to start admission server: %s\n", err)
+		os.Exit(1)
+	}
+}
+
+func startAdmissionServer() error {
+	ll := ctrl.Log.WithName("digitalocean-cloud-controller-manager-admission-server")
+
+	server := webhook.NewServer(webhook.Options{})
+
+	godoClient, err := getGodoClient()
+	if err != nil {
+		return fmt.Errorf("failed to get godo client: %s", err)
+	}
+
+	lbAdmissionHandler := do.NewLBServiceAdmissionHandler(&ll, godoClient)
+	if err := lbAdmissionHandler.WithRegion(); err != nil {
+		return fmt.Errorf("failed to inject region into lb service admission handler: %w", err)
+	}
+
+	clusterID := os.Getenv(doClusterIDEnv)
+	if clusterID == "" {
+		return fmt.Errorf("environment variable %q is required", doClusterIDEnv)
+	}
+	lbAdmissionHandler.WithClusterID(clusterID)
+
+	vpcID := os.Getenv(doClusterVPCIDEnv)
+	if vpcID == "" {
+		return fmt.Errorf("environment variable %q is required", doClusterVPCIDEnv)
+	}
+	lbAdmissionHandler.WithVPCID(vpcID)
+
+	ll.Info("registering admission handlers")
+	server.Register("/lb-service", &webhook.Admission{Handler: lbAdmissionHandler})
+
+	if err = server.Start(signals.SetupSignalHandler()); err != nil {
+		return err
+	}
+	ll.Info("Webhooks server started")
+
+	return nil
+}
+
+func getGodoClient() (*godo.Client, error) {
+	token := os.Getenv(doAccessTokenEnv)
```

Comment:
Env vars should ideally only be read/parsed at the top-most (main) function, and the values should be passed into functions.

We might have just refactored this from elsewhere though, in which case I guess we could keep it the way it is for now.

Reply:
I'll address when I'll work on metrics serving.